### PR TITLE
[Pipelined] Add support for static rules activation in gy app

### LINKED ISSUE #2719: [SessionD] Add FUA-Restrict Support
This project tracks the FUA-Restrict progress separately: https://github.com/magma/magma/projects/7#card-45998136

### PR DESCRIPTION
**Pull request by ymasmoudi**
Signed-off-by: YOUSSEF EL MASMOUDI <ymasmoudi@fb.com>

## Summary

This PR add the support for
1/ static rules activation in Gy Controller when final unit action is set to restrict_access.
2/ Qos Manager 
3/ Uses enforcement_stats table to retrieve statistics
4/ Add a unit test for restriction

## Test Plan

**CwF Integration Test including TestGyCreditExhaustionRestrict**

[vagrant@127.0.0.1:2201] out: 
[vagrant@127.0.0.1:2201] out: Running TestGyCreditExhaustionRestrict...
[vagrant@127.0.0.1:2201] out: ************************* TestRunner setup
[vagrant@127.0.0.1:2201] out: Adding Mock HSS service at 192.168.70.101:9204
[vagrant@127.0.0.1:2201] out: Adding Mock PCRF service at 192.168.70.101:9202
[vagrant@127.0.0.1:2201] out: Adding Mock OCS service at 192.168.70.101:9201
[vagrant@127.0.0.1:2201] out: Adding Pipelined service at 192.168.70.101:8443
[vagrant@127.0.0.1:2201] out: Adding Redis service at 192.168.70.101:6380
[vagrant@127.0.0.1:2201] out: Adding Directoryd service at 192.168.70.101:8443
[vagrant@127.0.0.1:2201] out: ************************* Configuring 1 UE(s), PCRF instance: PCRF_REMOTE
[vagrant@127.0.0.1:2201] out: Added UE to Simulator, HSS_REMOTE, PCRF_REMOTE, and OCS_REMOTE:
[vagrant@127.0.0.1:2201] out:   IMSI: 251441567100418   Key: a83ce3feb90ffc24ab1fb7f3df8a96e5   Opc: a3d29563f648a5e18f24b2fb37226b55   Seq: 2314075300928
[vagrant@127.0.0.1:2201] out: Successfully configured UE(s)
[vagrant@127.0.0.1:2201] out: ************************* Adding PCRF Usage Monitor for UE with IMSI: 251441567100418
[vagrant@127.0.0.1:2201] out: ************************* Adding a Pass-All static rule: static-pass-all-ocs1
[vagrant@127.0.0.1:2201] out: ************************* Adding a Pass-All static rule: static-pass-all-ocs2
[vagrant@127.0.0.1:2201] out: ************************* Adding PCRF Rule for UE with IMSI: 251441567100418 with ruleNames=[static-pass-all-ocs1 static-pass-all-ocs2], baseNames=[]
[vagrant@127.0.0.1:2201] out: ************************* Adding a static rule: restrict-deny-all
[vagrant@127.0.0.1:2201] out: ************************* Adding a static rule: restrict-pass-user
[vagrant@127.0.0.1:2201] out: ************************* Authenticating UE with IMSI: 251441567100418
[vagrant@127.0.0.1:2201] out: Finished Authenticating UE
[vagrant@127.0.0.1:2201] out: ************************* Generating Traffic for UE with Req: imsi:"251441567100418" volume:<value:"4.5M" > 
[vagrant@127.0.0.1:2201] out: Record sid:"IMSI251441567100418" rule_id:"restrict-deny-all" ue_ipv4:"192.168.128.2" 
[vagrant@127.0.0.1:2201] out: Record sid:"IMSI251441567100418" rule_id:"static-pass-all-ocs1" 
[vagrant@127.0.0.1:2201] out: Record sid:"IMSI251441567100418" rule_id:"static-pass-all-ocs2" bytes_tx:4564573 bytes_rx:26999 
[vagrant@127.0.0.1:2201] out: Record sid:"IMSI251441567100418" rule_id:"restrict-pass-user" bytes_rx:1500 ue_ipv4:"192.168.128.2" 
[vagrant@127.0.0.1:2201] out: ************************* Generating Traffic for UE with Req: imsi:"251441567100418" volume:<value:"2M" > 
[vagrant@127.0.0.1:2201] out: Record sid:"IMSI251441567100418" rule_id:"restrict-deny-all" ue_ipv4:"192.168.128.2" 
[vagrant@127.0.0.1:2201] out: Record sid:"IMSI251441567100418" rule_id:"static-pass-all-ocs1" 
[vagrant@127.0.0.1:2201] out: Record sid:"IMSI251441567100418" rule_id:"static-pass-all-ocs2" bytes_tx:4564573 bytes_rx:26999 
[vagrant@127.0.0.1:2201] out: Record sid:"IMSI251441567100418" rule_id:"restrict-pass-user" bytes_tx:2000110 bytes_rx:25647 ue_ipv4:"192.168.128.2" 
[vagrant@127.0.0.1:2201] out: Asserting all Gy expectations were met...
[vagrant@127.0.0.1:2201] out: Passed!
[vagrant@127.0.0.1:2201] out: ************************* Generating Traffic for UE with Req: imsi:"251441567100418" volume:<value:"2M" > 
[vagrant@127.0.0.1:2201] out: Record sid:"IMSI251441567100418" rule_id:"static-pass-all-ocs2" bytes_tx:2065458 bytes_rx:21519 
[vagrant@127.0.0.1:2201] out: --- PASS: TestGyCreditExhaustionRestrict (35.42s)
[vagrant@127.0.0.1:2201] out: PASS
[vagrant@127.0.0.1:2201] out: ok        magma/cwf/gateway/integ_tests   35.488s
[vagrant@127.0.0.1:2201] out: 
[vagrant@127.0.0.1:2201] out: DONE 1 tests in 42.129s
[vagrant@127.0.0.1:2201] out: 

**Unit Test**
................RTNETLINK answers: No such file or directory
We have an error talking to the kernel
RTNETLINK answers: No such file or directory
RTNETLINK answers: No such file or directory
We have an error talking to the kernel
RTNETLINK answers: No such file or directory
..QueryMatch(pkts=0, flow_count=1)
FlowTest(query=<magma.pipelined.tests.app.flow_query.RyuDirectFlowQuery object at 0x7f4ecad930b8>, match_num=0, flow_count=1)
QueryMatch(pkts=0, flow_count=1)
FlowTest(query=<magma.pipelined.tests.app.flow_query.RyuDirectFlowQuery object at 0x7f4ecad93400>, match_num=0, flow_count=1)
QueryMatch(pkts=0, flow_count=1)
FlowTest(query=<magma.pipelined.tests.app.flow_query.RyuDirectFlowQuery object at 0x7f4ecad93208>, match_num=0, flow_count=1)
.QueryMatch(pkts=0, flow_count=1)
FlowTest(query=<magma.pipelined.tests.app.flow_query.RyuDirectFlowQuery object at 0x7f4ecad9f710>, match_num=0, flow_count=1)
QueryMatch(pkts=0, flow_count=1)
FlowTest(query=<magma.pipelined.tests.app.flow_query.RyuDirectFlowQuery object at 0x7f4ecad9ff28>, match_num=0, flow_count=1)
QueryMatch(pkts=0, flow_count=1)
FlowTest(query=<magma.pipelined.tests.app.flow_query.RyuDirectFlowQuery object at 0x7f4ecad9fd30>, match_num=0, flow_count=1)
QueryMatch(pkts=0, flow_count=1)
FlowTest(query=<magma.pipelined.tests.app.flow_query.RyuDirectFlowQuery object at 0x7f4ecad9ff60>, match_num=0, flow_count=1)
QueryMatch(pkts=0, flow_count=1)
FlowTest(query=<magma.pipelined.tests.app.flow_query.RyuDirectFlowQuery object at 0x7f4ecad9fac8>, match_num=0, flow_count=1)
....QueryMatch(pkts=0, flow_count=1)
FlowTest(query=<magma.pipelined.tests.app.flow_query.RyuDirectFlowQuery object at 0x7f4ecabfd748>, match_num=0, flow_count=1)
QueryMatch(pkts=0, flow_count=1)
FlowTest(query=<magma.pipelined.tests.app.flow_query.RyuDirectFlowQuery object at 0x7f4ecabfd048>, match_num=0, flow_count=1)
QueryMatch(pkts=0, flow_count=1)
FlowTest(query=<magma.pipelined.tests.app.flow_query.RyuDirectFlowQuery object at 0x7f4ecabfdc18>, match_num=0, flow_count=1)
QueryMatch(pkts=0, flow_count=1)
FlowTest(query=<magma.pipelined.tests.app.flow_query.RyuDirectFlowQuery object at 0x7f4ecabfda58>, match_num=0, flow_count=1)
QueryMatch(pkts=0, flow_count=1)
FlowTest(query=<magma.pipelined.tests.app.flow_query.RyuDirectFlowQuery object at 0x7f4ecabfdba8>, match_num=0, flow_count=1)
...OrderedDict([('mme', <magma.pipelined.service_manager.Tables object at 0x7f4ecabba548>), ('ingress', <magma.pipelined.service_manager.Tables object at 0x7f4ecaba69c8>), ('arpd', <magma.pipelined.service_manager.Tables object at 0x7f4ecabba348>), ('access_control', <magma.pipelined.service_manager.Tables object at 0x7f4ecabba388>), ('middle', <magma.pipelined.service_manager.Tables object at 0x7f4ecaba6a08>), ('dpi', <magma.pipelined.service_manager.Tables object at 0x7f4ecabba3c8>), ('gy', <magma.pipelined.service_manager.Tables object at 0x7f4ecabba408>), ('enforcement', <magma.pipelined.service_manager.Tables object at 0x7f4ecabba448>), ('enforcement_stats', <magma.pipelined.service_manager.Tables object at 0x7f4ecabba488>), ('ipfix', <magma.pipelined.service_manager.Tables object at 0x7f4ecabba4c8>), ('egress', <magma.pipelined.service_manager.Tables object at 0x7f4ecaba6a48>)])
......QueryMatch(pkts=2, flow_count=5)
FlowTest(query=<magma.pipelined.tests.app.flow_query.RyuDirectFlowQuery object at 0x7f4ecab18dd8>, match_num=2, flow_count=5)
QueryMatch(pkts=1, flow_count=1)
FlowTest(query=<magma.pipelined.tests.app.flow_query.RyuDirectFlowQuery object at 0x7f4ecab182b0>, match_num=1, flow_count=1)
QueryMatch(pkts=1, flow_count=1)
FlowTest(query=<magma.pipelined.tests.app.flow_query.RyuDirectFlowQuery object at 0x7f4ecab18b38>, match_num=1, flow_count=1)
.QueryMatch(pkts=2, flow_count=3)
FlowTest(query=<magma.pipelined.tests.app.flow_query.RyuDirectFlowQuery object at 0x7f4ecab18898>, match_num=1, flow_count=3)
QueryMatch(pkts=0, flow_count=0)
FlowTest(query=<magma.pipelined.tests.app.flow_query.RyuDirectFlowQuery object at 0x7f4ecab18b38>, match_num=0, flow_count=0)
QueryMatch(pkts=2, flow_count=1)
FlowTest(query=<magma.pipelined.tests.app.flow_query.RyuDirectFlowQuery object at 0x7f4ecab18d30>, match_num=1, flow_count=1)
.QueryMatch(pkts=4, flow_count=3)
FlowTest(query=<magma.pipelined.tests.app.flow_query.RyuDirectFlowQuery object at 0x7f4ecab29080>, match_num=4, flow_count=3)
QueryMatch(pkts=4, flow_count=2)
FlowTest(query=<magma.pipelined.tests.app.flow_query.RyuDirectFlowQuery object at 0x7f4ecab29e10>, match_num=4, flow_count=2)
QueryMatch(pkts=3, flow_count=2)
FlowTest(query=<magma.pipelined.tests.app.flow_query.RyuDirectFlowQuery object at 0x7f4ecab29da0>, match_num=3, flow_count=2)
QueryMatch(pkts=4, flow_count=1)
FlowTest(query=<magma.pipelined.tests.app.flow_query.RyuDirectFlowQuery object at 0x7f4ecab29630>, match_num=4, flow_count=1)
.ifup: interface eth3 already configured
.ovs-vsctl: bridge upt_br0 does not have a port eth_t1
.RTNETLINK answers: File exists
.RTNETLINK answers: File exists
RTNETLINK answers: File exists
.RTNETLINK answers: File exists
..

### CODE REVIEW

File: lte/gateway/python/magma/pipelined/app/gy.py [hunk 1]
```diff
@@ -37,25 +48,35 @@ class GYController(PolicyMixin, MagmaController):
...
+    # Effectively range is 3 -> 65535
+    GY_PRIORITY_RANGE = MAX_GY_PRIORITY - MIN_GY_PROGRAMMED_FLOW
 
     def __init__(self, *args, **kwargs):
         super(GYController, self).__init__(*args, **kwargs)
+        self._config = kwargs['config']
         self.tbl_num = self._service_manager.get_table_num(self.APP_NAME)
         self.next_main_table = self._service_manager.get_next_table_num(
             self.APP_NAME)
+        self._enforcement_stats_scratch = self._service_manager.get_table_num(
```
Comment (koolzz) [inline lte/gateway/python/magma/pipelined/app/gy.py]: `_enforcement_stats_scratch` -> `_enforcement_stats_tbl`

Its actually a regular table so lets use proper naming, I noticed that I forgot to refactor enf controller(I'll refactor it there)

File: lte/gateway/python/magma/pipelined/app/gy.py [hunk 2]
```diff
@@ -229,6 +294,185 @@ def _install_default_flows_if_not_installed(self, datapath,
...
+                    rule.app_service_type))
+
+            except FlowMatchError as err:  # invalid match
+                self.logger.error(
+                    "Failed to get flow msg '%s' for subscriber %s: %s",
+                    rule.id, imsi, err)
+                raise err
+        return flow_adds
+
+    def _get_classify_rule_flow_msgs(self, imsi, ip_addr, apn_ambr, flow, rule_num,
```
Comment (koolzz) [inline lte/gateway/python/magma/pipelined/app/gy.py]: This seems like a lot of code duplication, is enforcement stats doing the exact same thing? Could we add the next table as an arg and put this into a shared file?

File: lte/gateway/python/magma/pipelined/app/gy.py [hunk 3]
```diff
@@ -229,6 +294,185 @@ def _install_default_flows_if_not_installed(self, datapath,
...
+                flow_match_actions,
+                instructions=instructions,
+                hard_timeout=hard_timeout,
+                priority=priority,
+                cookie=rule_num,
+                resubmit_table=self._enforcement_stats_scratch)
+            )
+        return msgs
+
+    def _get_classify_rule_of_actions(self, flow, rule_num, imsi, ip_addr,
```
Comment (koolzz) [inline lte/gateway/python/magma/pipelined/app/gy.py]: Same for this, can we just put qos manager and version mapper as args and move to a common file?

File: lte/gateway/python/magma/pipelined/app/gy.py [hunk 4]
```diff
@@ -229,6 +294,185 @@ def _install_default_flows_if_not_installed(self, datapath,
...
+
+        version = self._session_rule_version_mapper.get_version(imsi, ip_addr,
+                                                                rule_id)
+        actions.extend(
+            [parser.NXActionRegLoad2(dst='reg2', value=rule_num),
+             parser.NXActionRegLoad2(dst=RULE_VERSION_REG, value=version)
+             ])
+        return actions, instructions
+
+    def get_of_priority(self, precedence):
```
Comment (koolzz) [inline lte/gateway/python/magma/pipelined/app/gy.py]: Same idea, I think all the priorities are the same? So just make them common

**Review by koolzz — CHANGES_REQUESTED** [review]
lets try to avoid so much code dup

File: lte/gateway/python/magma/pipelined/app/gy.py (hunk 2 above)
Comment (ymasmoudi) [reply lte/gateway/python/magma/pipelined/app/gy.py]: Yes, this was left in purpose as I don't have visibility on the planned work on the enforcement side and whether we want the apps code to be decoupled or not.
all the code is factored now.

**Review by koolzz — APPROVED** [review]
Overall lgtm, hold off on merging before @pshelar takes a look 

Also do you want to include the restrict + enforcement stats rule unit test?

File: lte/gateway/python/magma/pipelined/app/policy_mixin.py [hunk 5]
```diff
@@ -239,6 +273,136 @@ def fail(err):
...
+                instructions=instructions,
+                hard_timeout=hard_timeout,
+                priority=priority,
+                cookie=rule_num,
+                resubmit_table=next_table)
+            )
+        return msgs
+
+    def _get_classify_rule_of_actions(self, flow, rule_num, imsi, ip_addr,
+                                      apn_ambr, qos, rule_id, version, qos_mgr):
```
Comment (pshelar) [inline lte/gateway/python/magma/pipelined/app/policy_mixin.py]: The name is bit confusing, can we call it _get_action_for_rule()

File: lte/gateway/python/magma/pipelined/tests/test_gy.py [hunk 6]
```diff
@@ -160,6 +166,63 @@ def test_subscriber_policy(self):
...
             pkt_sender.send(packet)
 
+    def test_subscriber_restrict_policy(self):
+        """
+        Add policy to subscriber, send 4096 packets
+
+        Assert:
+            Packets are properly matched with the 'simple_match' policy
+            Send /20 (4096) packets, match /16 (256) packets
+        """
```
Comment (pshelar) [inline lte/gateway/python/magma/pipelined/tests/test_gy.py:177]: update the desc here for test

File: lte/gateway/python/magma/pipelined/tests/test_gy.py [hunk 7]
```diff
@@ -160,6 +166,63 @@ def test_subscriber_policy(self):
...
+            imsi, sub_ip, self.gy_controller, self._tbl_num
+        ).add_static_rule(policies[0].id)
+        isolator = RyuDirectTableIsolator(
+            RyuForwardFlowArgsBuilder.from_subscriber(sub_context.cfg)
+                                     .build_requests(),
+            self.testing_controller
+        )
+        pkt_sender = ScapyPacketInjector(self.IFACE)
+        packet = IPPacketBuilder()\
+            .set_ip_layer('45.10.0.0/20', sub_ip)\
```
Comment (pshelar) [inline lte/gateway/python/magma/pipelined/tests/test_gy.py]: cidr for IP address in packet looks strange, lets just keep the IP address here.

**Review by pshelar — APPROVED** [review]
I have couple of comments, fix those before merge.

**Pull request merged**